File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/offline/OfflineSettlementBuildClaimOverride.java
package org.hyperledger.iroha.android.offline;

import java.util.LinkedHashMap;
import java.util.Map;

/** Optional per-receipt build-claim override for offline settlement submission. */
public final class OfflineSettlementBuildClaimOverride {

  private final String txIdHex;
  private final String appId;
  private final Long buildNumber;
  private final Long issuedAtMs;
  private final Long expiresAtMs;

  private OfflineSettlementBuildClaimOverride(final Builder builder) {
    this.txIdHex =
        OfflineHashLiteral.parseHex(requireNonBlank(builder.txIdHex, "txIdHex"), "txIdHex");
    this.appId = normalizeOptional(builder.appId);
    this.buildNumber = requireNonNegative(builder.buildNumber, "buildNumber");
    this.issuedAtMs = requireNonNegative(builder.issuedAtMs, "issuedAtMs");
    this.expiresAtMs = requireNonNegative(builder.expiresAtMs, "expiresAtMs");
  }

  public static Builder builder() {
    return new Builder();
  }

  public Map<String, Object> toJsonMap() {
    final Map<String, Object> json = new LinkedHashMap<>();
    json.put("tx_id_hex", txIdHex);
    if (appId != null) {
      json.put("app_id", appId);
    }
    if (buildNumber != null) {
      json.put("build_number", buildNumber);
    }
    if (issuedAtMs != null) {
      json.put("issued_at_ms", issuedAtMs);
    }
    if (expiresAtMs != null) {
      json.put("expires_at_ms", expiresAtMs);
    }
    return json;
  }

  private static String requireNonBlank(final String value, final String name) {
    final String normalized = normalizeOptional(value);
    if (normalized == null) {
      throw new IllegalArgumentException(name + " must be provided");
    }
    return normalized;
  }

  private static String normalizeOptional(final String value) {
    if (value == null) {
      return null;
    }
    final String trimmed = value.trim();
    return trimmed.isEmpty() ? null : trimmed;
  }

  private static Long requireNonNegative(final Long value, final String name) {
    if (value != null && value < 0) {
      throw new IllegalArgumentException(name + " must not be negative");
    }
    return value;
  }

  public static final class Builder {
    private String txIdHex;
    private String appId;
    private Long buildNumber;
    private Long issuedAtMs;
    private Long expiresAtMs;

    private Builder() {}

    public Builder txIdHex(final String txIdHex) {
      this.txIdHex = txIdHex;
      return this;
    }

    public Builder appId(final String appId) {
      this.appId = appId;
      return this;
    }

    public Builder buildNumber(final Long buildNumber) {
      this.buildNumber = buildNumber;
      return this;
    }

    public Builder issuedAtMs(final Long issuedAtMs) {
      this.issuedAtMs = issuedAtMs;
      return this;
    }

    public Builder expiresAtMs(final Long expiresAtMs) {
      this.expiresAtMs = expiresAtMs;
      return this;
    }

    public OfflineSettlementBuildClaimOverride build() {
      return new OfflineSettlementBuildClaimOverride(this);
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/offline/OfflineSettlementBuildClaimOverride.kt
package org.hyperledger.iroha.sdk.offline

/** Optional per-receipt build-claim override for offline settlement submission. */
class OfflineSettlementBuildClaimOverride private constructor(
    private val txIdHex: String,
    private val appId: String?,
    private val buildNumber: Long?,
    private val issuedAtMs: Long?,
    private val expiresAtMs: Long?,
) {
    fun toJsonMap(): Map<String, Any> {
        val json = LinkedHashMap<String, Any>()
        json["tx_id_hex"] = txIdHex
        if (appId != null) json["app_id"] = appId
        if (buildNumber != null) json["build_number"] = buildNumber
        if (issuedAtMs != null) json["issued_at_ms"] = issuedAtMs
        if (expiresAtMs != null) json["expires_at_ms"] = expiresAtMs
        return json
    }

    class Builder {
        private var txIdHex: String? = null
        private var appId: String? = null
        private var buildNumber: Long? = null
        private var issuedAtMs: Long? = null
        private var expiresAtMs: Long? = null

        fun txIdHex(txIdHex: String?) = apply { this.txIdHex = txIdHex }
        fun appId(appId: String?) = apply { this.appId = appId }
        fun buildNumber(buildNumber: Long?) = apply { this.buildNumber = buildNumber }
        fun issuedAtMs(issuedAtMs: Long?) = apply { this.issuedAtMs = issuedAtMs }
        fun expiresAtMs(expiresAtMs: Long?) = apply { this.expiresAtMs = expiresAtMs }

        fun build(): OfflineSettlementBuildClaimOverride {
            val normalizedTxIdHex = requireNotNull(txIdHex?.trim()?.ifEmpty { null }) {
                "txIdHex must be provided"
            }
            val validatedTxIdHex = OfflineHashLiteral.parseHex(normalizedTxIdHex, "txIdHex")
            val normalizedAppId = appId?.trim()?.ifEmpty { null }
            require(buildNumber == null || buildNumber!! >= 0) { "buildNumber must not be negative" }
            require(issuedAtMs == null || issuedAtMs!! >= 0) { "issuedAtMs must not be negative" }
            require(expiresAtMs == null || expiresAtMs!! >= 0) { "expiresAtMs must not be negative" }
            return OfflineSettlementBuildClaimOverride(
                validatedTxIdHex, normalizedAppId, buildNumber, issuedAtMs, expiresAtMs
            )
        }
    }

    companion object {
        @JvmStatic
        fun builder() = Builder()
    }
}
